beacon_node/beacon_chain/src/effective_balances_cache.rs

```diff
@@ -0,0 +1,179 @@
+use crate::{BeaconChain, BeaconChainError, BeaconChainTypes};
+use parking_lot::{RwLock, RwLockUpgradableReadGuard};
+use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
+use types::{BeaconCommittee, BeaconState, Epoch, EthSpec, ForkName, Slot};
+
+// Max number of epochs this cache will store data for
+pub const EFFECTIVE_BALANCE_CACHE_SIZE: usize = 8;
+
+#[derive(Debug)]
+pub enum Error {
+    StateUnavailable(Epoch),
+    CacheMiss(Epoch),
+    ValidatorIndexUnknown(usize),
+    EpochOutOfBounds(Epoch),
+}
+
+#[derive(Clone)]
+pub struct EffectiveBalancesCache {
+    pub effective_balances: Arc<RwLock<BTreeMap<Epoch, Arc<Vec<u64>>>>>,
```

Comment:
You need to get the balances of the specific fork of this attestation, in highly divergent networks just indexing by epoch is not enough